mobile_back_apple/cpp/backend_coreml/main.cc

```diff
@@ -0,0 +1,131 @@
+  return ((CoreMLBackendData *)backend_ptr)->vendor;
+}
+
+// Return the name of the accelerator.
+const char *mlperf_backend_accelerator_name(mlperf_backend_ptr_t backend_ptr) {
+  return ((CoreMLBackendData *)backend_ptr)->accelerator;
+}
+
+// Return the name of this backend.
+const char *mlperf_backend_name(mlperf_backend_ptr_t backend_ptr) {
+  return ((CoreMLBackendData *)backend_ptr)->name;
+}
+
+// Should return true if current hardware is supported.
+bool mlperf_backend_matches_hardware(const char **not_allowed_message,
+                                     const char **settings,
+                                     const mlperf_device_info_t *device_info) {
+  (void)device_info;
+  *settings = coreml_settings.c_str();
+  *not_allowed_message = "This is a stub backend for Core ML";
+  return true;
+}
+
+// Create a new backend and return the pointer to it.
+mlperf_backend_ptr_t mlperf_backend_create(
+    const char *model_path, mlperf_backend_configuration_t *configs,
+    const char *native_lib_path) {
+  // Verify only one instance of the backend exists at any time
+  if (backendExists) {
+    LOG(INFO) << "Only one backend instance should exist at a time";
+    return nullptr;
+  }
+
+  CoreMLBackendData *backend_data = new CoreMLBackendData();
+
+  backendExists = true;
+
+  // Load the model.
+  CoreMLExecutor *coreMLExecutor = [[CoreMLExecutor alloc] initWithModelPath:model_path];
+  if (!coreMLExecutor) {
+    LOG(INFO) << "Cannot create CoreMLExecutor";
+    return nullptr;
+  }
+  LOG(INFO) << "CoreMLExecutor created";
+  backend_data->coreMLExecutor = coreMLExecutor;
+  return backend_data;
+}
+
+// Destroy the backend pointer and its data.
+void mlperf_backend_delete(mlperf_backend_ptr_t backend_ptr) {
+  (void)backend_ptr;
+}
+
+// Run the inference for a sample.
+mlperf_status_t mlperf_backend_issue_query(mlperf_backend_ptr_t backend_ptr) {
+  LOG(INFO) << "mlperf_backend_issue_query()";
+  return MLPERF_FAILURE;
+}
+
+// Flush the staged queries immediately.
+mlperf_status_t mlperf_backend_flush_queries(mlperf_backend_ptr_t backend_ptr) {
+  LOG(INFO) << "mlperf_backend_flush_queries()";
+  return MLPERF_FAILURE;
+}
+
+// Return the number of inputs of the model.
+int32_t mlperf_backend_get_input_count(mlperf_backend_ptr_t backend_ptr) {
+  return [((CoreMLBackendData *) backend_ptr)->coreMLExecutor getInputCount];
+}
+
+// Return the type of the ith input.
+mlperf_data_t mlperf_backend_get_input_type(mlperf_backend_ptr_t backend_ptr,
+                                            int32_t i) {
+  enum mlperf_data_t::Type datatype = mlperf_data_t::Type::Float32;
+  mlperf_data_t data = {.type = datatype, .size = 0};
+  return data;
+}
+
+// Set the data for ith input.
+mlperf_status_t mlperf_backend_set_input(mlperf_backend_ptr_t backend_ptr,
+                                         int32_t batchIndex, int32_t i,
+                                         void *data) {
+  LOG(INFO) << "mlperf_backend_set_input()";
```

Comment:
@freedomtan I use your code example from TestSimpleCoreMLCommandLine to implement some methods for CoreMLExecutor. It works fine when tested it directly using CLI, so it should not be related to the following error.

When running `make coreml` I got this error
```
2022-07-13 11:21:25.090745: I mobile_back_apple/cpp/backend_coreml/main.cc:67] CoreMLExecutor created
2022-07-13 11:21:25.090 main[84821:15177994] inputCount 1
2022-07-13 11:21:25.090 main[84821:15177994] outputCount 1
2022-07-13 11:21:25.090778: I flutter/cpp/binary/main.cc:196] Using Imagenet dataset
libc++abi: terminating with uncaught exception of type std::length_error: allocator<T>::allocate(size_t n) 'n' exceeds maximum supported size
```

I expected the functions will be called in this order:

1. mlperf_backend_create()
2. mlperf_backend_get_input_count()
3. mlperf_backend_get_input_type()
4. mlperf_backend_get_output_count()
5. mlperf_backend_get_output_type()
6. mlperf_backend_set_input()
7. mlperf_backend_issue_query()
8. mlperf_backend_get_output()
9. mlperf_backend_flush_queries()

So it looks like the error is occurred before `mlperf_backend_set_input()` get called. Do you have an idea why this happens?

Reply:
@anhappdev you didn't set the input and output types correctly. Sizes for float32 should 4.

Reply:
Ah ok, thanks.